waiverdb/models/requests.py

```diff
@@ -0,0 +1,117 @@
+from datetime import datetime
+
+import pydantic
+from pydantic import BaseModel, Field
+from werkzeug.exceptions import BadRequest
+
+
+# WaiverDB < 0.11 compatibility
+class TestSubject(BaseModel):
+    type: Optional[str]
+    item: Optional[str]
+    original_spec_nvr: Optional[str]
+    productmd_compose_id: Optional[str] = Field(alias='productmd.compose.id', default=None)
+
+
+class TestResult(BaseModel):
+    testcase: str
+    subject: TestSubject
+
+
+class CreateWaiver(BaseModel):
+    subject_type: Optional[str]
+    subject_identifier: Optional[str]
+    testcase: Optional[str]
+    subject: Optional[TestSubject]
+    result_id: Optional[int]
+    waived: bool = True
+    product_version: str
+    comment: str
+    username: Optional[str] = None
+    scenario: Optional[str] = None
+
+
+class GetWaivers(BaseModel):
+    subject_type: Optional[str]
+    subject_identifier: Optional[str]
+    testcase: Optional[str]
+    product_version: Optional[str]
+    username: Optional[str]
+    include_obsolete: bool = False
+    scenario: Optional[str] = None
+    since: Optional[str]
+    page: int = 1
+    limit: int = 10
+    proxied_by: Optional[str]
+
+
+class GetPermissions(BaseModel):
+    testcase: Optional[str]
+
+
+class WaiverFilter(BaseModel):
+    subject_type: Optional[str]
+    subject_identifier: Optional[str]
+    testcase: Optional[str]
+    scenario: Optional[str]
+    product_version: Optional[str]
+    username: Optional[str]
+    proxied_by: Optional[str]
+    since: Optional[str]
+
+
+class FilterWaivers(BaseModel):
+    filters: List[WaiverFilter]
+    include_obsolete: bool = False
+
+
+class GetWaiversBySubjectAndTestcase(BaseModel):
+    results: Optional[List[TestResult]]
+    testcase: Optional[str]
+    product_version: Optional[str]
+    username: Optional[str]
+    proxied_by: Optional[str]
+    since: Optional[str]
+    include_obsolete: bool = False
+
+
+def parse_since(since: str) -> Tuple[Optional[datetime], Optional[datetime]]:
+    """
+    Parses the 'since' query parameter, which is expected to be either a
+    single ISO8601 timestamp representing the start of a time period::
+
+        2017-02-13T23:37:58.193281
+
+    or a comma-separated pair of timestamps representing the start and end of
+    a range::
+
+        2017-02-13T23:37:58.193281,2017-02-16T23:37:58.193281
+
+    Returns a tuple (start, end) of datetime.datetime instances.
+    """
+    start = None
+    end = None
+    if ',' in since:
+        start, end = since.split(',', 1)
+    else:
+        start = since
+    try:
+        if start:
+            start = datetime.fromisoformat(start)
+        if end:
+            end = datetime.fromisoformat(end)
+    except ValueError as e:
+        raise BadRequest({'since': str(e)})
+    return start, end
+
+
+def parse_validation_error(err: pydantic.ValidationError) -> Dict[str, str]:
+    # TODO: this is here to make errors compatible with previous,
+    # we should probably refactor this later
```

Comment:
We do not need to keep the error format compatible with previous versions. It is not part of API though I can see it can break some behavior.

One thing that would break waiverdb-cli (even though it should not be used anymore) is that the "message" field may not be available in the response for failure when creating waiver.

Reply:
For me it seems very useful not only because of compatibility, but it also a good way to jsonify validation errors, so instead of pure exception and 500 error user can get better information regarding what's wrong with his response.